src/translator/translation_model.h

```diff
@@ -0,0 +1,122 @@
+      : TranslationModel(parseOptionsFromString(config, /*validate=*/false), std::move(memory), replicas){};
+
+  /// Construct TranslationModel from marian-options. If memory is empty, TranslationModel is initialized from
+  /// paths available in the options object, backed by filesystem. Otherwise, TranslationModel is initialized from the
+  /// given MemoryBundle composed of AlignedMemory holding equivalent parameters.
+  ///
+  /// @param [in] options: Marian options object.
+  /// @param [in] memory: MemoryBundle object holding memory buffers containing parameters to build MarianBackend,
+  /// ShortlistGenerator, Vocabs and SentenceSplitter.
+  TranslationModel(const Config& options, MemoryBundle&& memory = MemoryBundle{}, size_t replicas = 1);
+
+  /// Make a Request to be translated by this TranslationModel instance.
+  /// @param [in] requestId: Unique identifier associated with this request, available from Service.
+  /// @param [in] source: Source text to be translated. Ownership is accepted and eventually returned to the client in
+  /// Response corresponding to the Request created here.
+  /// @param [in] callback: Callback (from client) to be issued upon completion of translation of all sentences in the
+  /// created Request.
+  /// @param [in] responseOptions: Configuration used to prepare the Response corresponding to the created request.
+  //  @returns Request created from the query parameters wrapped within a shared-pointer.
+  Ptr<Request> makeRequest(size_t requestId, std::string&& source, CallbackType callback,
+                           const ResponseOptions& responseOptions);
+
+  /// Relays a request to the batching-pool specific to this translation model.
+  /// @param [in] request: Request constructed through makeRequest
+  void enqueueRequest(Ptr<Request> request) { batchingPool_.enqueueRequest(request); };
+
+  /// Generates a batch from the batching-pool for this translation model, compiling from several active requests. Note
+  /// that it is possible that calls to this method can give empty-batches.
+  ///
+  /// @param [out] batch: Batch to write a generated batch on to.
+  /// @returns number of sentences that constitute the Batch.
+  size_t generateBatch(Batch& batch) { return batchingPool_.generateBatch(batch); }
+
+  /// Translate a batch generated with generateBatch
+  ///
+  /// @param [in] deviceId: There are replicas of backend created for use in each worker thread. deviceId indicates
+  /// which replica to use.
+  /// @param [in] batch: A batch generated from generateBatch from the same TranslationModel instance.
+  void translateBatch(size_t deviceId, Batch& batch);
+
+ private:
+  Config options_;
+  MemoryBundle memory_;
+  Vocabs vocabs_;
+  TextProcessor textProcessor_;
+
+  /// Maintains sentences from multiple requests bucketed by length and sorted by priority in each bucket.
+  BatchingPool batchingPool_;
```

Comment:
It's weird to have a batching pool inside the TranslationModel.  What you could have had is something like a TranslationModel object without a pool, then a larger struct that has BatchingPool and TranslationModel as member variables.  But then I get that you'd have to look up the BatchingPool.  So from a user perspective I guess it's right to have them together.  Maybe slough off a class that just has the model stuff and make it a member variable here?

Reply:
In the current setting, providing the worker specifics or how-to translate, one can use API provided by `TranslationModel` to achieve translation of a request. The `BatchingPool` is a part of a given `TranslationModel` instance, aimed at generating homogeneous batches collating from multiple requests. Ownership is with `TranslationModel` as it is not possible to translate without the graph and other preproc existing. I perceive `TranslationModel` as "all things needed to translate in a single-direction", which includes the `BatchingPool`. 

I'm confused to how to separate `TranslationModel = ActualTranslationModel + BatchingPool` into meaningful units. `BatchingPool` handles batching (raw sentences in, optimally* packed batches out), which is understandable and keeping this concern separate makes sense. But regarding the other half - what is `ActualTranslationModel` intended to do? I'd need to know this to create a proper separation. It's never used by the client. It's one more layer deeper for library use (library works with `TranslationModel`).

Reply:
Carrying this outside this PR, if this needs to be addressed. Since it does not affect anything external, we can always come back and change this... so.